### DISCUSSION TITLE
Symbolic regression: Mixing Bool and Double Ops

### DISCUSSION THREAD

**Discussion by raedma** [jenetics/jenetics#848] [Q&A] [answered]
Hello, 

I am currently facing a problem, where I would like to find a symbolic regression for a function with discontinuities, e.g.

![image](https://user-images.githubusercontent.com/25957103/201039655-8d31e20c-d642-42ab-9c1e-42b35edc87c8.png)

I tried the example in [HeuristicLab](https://dev.heuristiclab.com/trac.fcgi/) and it gave me a solution with basically a conditional "if x smaller than or of x greater then":

![image](https://user-images.githubusercontent.com/25957103/201041044-9163cbf2-f1ef-4faa-b686-4658d781d4f9.png)

Is this possible with jenetics? At a first look I found [this](https://github.com/jenetics/jenetics/blob/master/jenetics.prog/src/test/java/io/jenetics/prog/regression/RegressionTest.java) nice example for a setup with doubles and have seen that [`BoolOp`](https://github.com/jenetics/jenetics/blob/master/jenetics.prog/src/main/java/io/jenetics/prog/op/BoolOp.java)s are supported. However, I was unable to find an example where both operator types are mixed to later build a `Regression`. `Regression` expects a generic type, in the example `Double`, but I guess `Boolean` would work just as well with `BoolOp`. Is it not possible to mix `Op<Double>` and `Op<Boolean>` to build an `ISeq<?>` to build a `Regression.codecOf(...)` or would it work just with the the use of `Object` as the generic type?

A small hint or example would be very nice.

Regards
Martin

**Comment by jenetics** [comment, marked as answer]
You can try an extension to the [MultiRootGP](https://github.com/jenetics/jenetics/blob/master/jenetics.example/src/main/java/io/jenetics/example/MultiRootGP.java) example.

```java
static final Codec<ISeq<Function<Double[], Double>>, ProgramGene<Double>> CODEC =
    Codec.of(
        Genotype.of(
            // Discriminator
            ProgramChromosome.of(
                0,
                ISeq.of(Op.of("identity", UnaryOperator.identity())),
                ISeq.of(EphemeralConst.of(() -> (double)RandomRegistry.random().nextInt(100)))
            ),
            // First 'program'
            ProgramChromosome.of(
                5,
                ch -> ch.root().size() <= 50,
                OPERATIONS,
                TERMINALS
            ),
            // Second 'program'
            ProgramChromosome.of(
                5,
                ch -> ch.root().size() <= 50,
                OPERATIONS,
                TERMINALS
            )
        ),
        gt -> gt.stream()
            .map(Chromosome::gene)
            .collect(ISeq.toISeq())
    );

static double fitness(final ISeq<Function<Double[], Double>> programs) {
    final var discriminator = programs.get(0).apply(new Double[] {0.0});
    if (discriminator <= 50) {
        // Evaluate programs.get(1)
        return 1;
    } else {
        // Evaluate program.get(2)
        return 2;
    }
}
```
